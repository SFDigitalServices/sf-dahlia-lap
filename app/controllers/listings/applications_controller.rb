# frozen_string_literal: true

module Listings
  # Rails controller for views/actions related to applications for a listing
  class ApplicationsController < ApplicationController
    before_action :authenticate_user!
    before_action :load_listing
    before_action :validate_listing!

    def index
      @applications = soql_application_service.listing_applications(params[:listing_id])
      @fields = soql_application_service.index_fields
    end

    def new
      @lending_institutions = listing_service.sale?(@listing) ? lending_institutions : {}
    end

    private

    def validate_listing!
      raise Force::RecordNotFound, "Listing #{params[:listing_id]} not found" unless @listing.present?
    end

    def load_listing
      @listing = listing_service.listing(params[:listing_id])
    end

    def listing_service
      Force::ListingService.new(current_user)
    end

    def soql_application_service
      Force::Soql::ApplicationService.new(current_user)
    end

    def lending_institutions
      Force::CustomApi::LendingInstitutionsService.new(current_user).lending_institutions
    end
  end
end
