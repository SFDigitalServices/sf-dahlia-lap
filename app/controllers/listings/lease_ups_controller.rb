# frozen_string_literal: true

module Listings
  # Rails controller for views/actions related to listings that have Lease Up status
  class LeaseUpsController < ApplicationController
    before_action :authenticate_user!

    def index
      @listings = soql_listing_service.lease_up_listings
    end

    private

    def soql_listing_service
      Force::Soql::ListingService.new(current_user)
    end
  end
end
