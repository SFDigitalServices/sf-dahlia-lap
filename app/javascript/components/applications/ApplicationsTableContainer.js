import React from 'react'

import { EagerPagination, SERVER_PAGE_SIZE } from 'utils/EagerPagination'

import ApplicationsFilter from './ApplicationsFilter'
import ApplicationsTable from './ApplicationsTable'

const ROWS_PER_PAGE = 20

class ApplicationsTableContainer extends React.Component {
  constructor(props) {
    super(props)
    this.eagerPagination = new EagerPagination(ROWS_PER_PAGE, SERVER_PAGE_SIZE)
    this.state = {
      filters: props.filters,
      loading: false,
      applications: [],
      pages: 0
    }
  }

  loadPage = (page, filters) => {
    const { onFetchData } = this.props
    const fetcher = (p) => onFetchData(p, { filters })
    this.setState({ loading: true, page: page })
    this.eagerPagination.getPage(page, fetcher).then(({ records, pages }) => {
      this.setState({ applications: records, loading: false, pages: pages, atMaxPages: false })
    })
  }

  handleOnFetchData = (state, instance) => {
    const { filters } = this.state
    if (this.eagerPagination.isOverLimit(state.page)) {
      this.setState({ applications: [], loading: false, atMaxPages: true })
    } else {
      this.loadPage(state.page, filters)
    }
  }

  handleOnFilter = (filters) => {
    this.setState({ filters })
    this.eagerPagination.reset()
    this.loadPage(0, filters)
  }

  render() {
    const { listings } = this.props
    const { loading, applications, pages, atMaxPages } = this.state
    return (
      <>
        <ApplicationsFilter onSubmit={this.handleOnFilter} listings={listings} loading={loading} />
        <ApplicationsTable
          applications={applications}
          onFetchData={this.handleOnFetchData}
          pages={pages}
          loading={loading}
          rowsPerPage={ROWS_PER_PAGE}
          atMaxPages={atMaxPages}
        />
      </>
    )
  }
}

export default ApplicationsTableContainer
