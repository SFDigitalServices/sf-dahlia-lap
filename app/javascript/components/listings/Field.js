import React from 'react'

var generateHtml = (value) => {
  return { __html: value }
}

const RenderType = ({ type, value }) => {
  if (type === 'html') {
    return <p dangerouslySetInnerHTML={generateHtml(value)} />
  } else if (type === 'link') {
    return (
      <a target='_blank' rel='noreferrer' href={value}>
        {value}
      </a>
    )
  } else {
    return <p>{String(value)}</p>
  }
}

const Content = ({ label, value, type }) => {
  return (
    <div className='margin-bottom--half'>
      <h4 className='t-sans t-small t-bold no-margin'>{label}</h4>
      <RenderType type={type} value={value} />
    </div>
  )
}

export default Content
