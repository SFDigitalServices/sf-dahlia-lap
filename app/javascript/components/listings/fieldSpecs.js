import { get, includes } from 'lodash'
import moment from 'moment'

import utils from '~/utils/utils'

const getRenderType = (value) => {
  if (includes(value, 'http')) {
    return 'link'
  } else {
    return null
  }
}

const formatValue = (value, type) => {
  if (type === 'date') {
    return moment(value).format('L')
  } else {
    return value
  }
}

export const buildFieldEntry = (item, entry) => {
  let value = get(item, entry.field)
  const label = utils.cleanField(entry.label)
  const renderType = entry.renderType || getRenderType(value)

  value = formatValue(value, entry.formatType)

  return { label, value, renderType }
}
