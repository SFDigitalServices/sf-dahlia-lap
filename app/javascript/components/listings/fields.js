export const detailsFields = [
  { field: 'owner', label: 'Owner' },
  'name',
  { field: 'account', label: 'Account' },
  'application_due_date',
  { field: 'in_lottery', label: 'Applications in Lottery' },
  'status',
  'lottery_winners',
  'lottery_results'
]

export const buildingInformationFields = [
  'building_name',
  'building_street_address',
  'building_city',
  'building_state',
  'building_zip_code',
  'neighborhood',
  'developer',
  { field: 'building_url', label: 'Building URL' },
  'year_built',
  'description',
  'lottery_preferences'
]

export const lotteryPreferencesFields = [
  { field: 'lottery_preference.name', label: 'Name' },
  'description',
  { field: 'pdf_url', label: 'PDF URL' },
  'order',
  'available_units'
]

export const aafFields = [
  'accessibility',
  'fee',
  'amenities',
  { field: 'deposit_min', formatType: 'currency' },
  { field: 'deposit_max', formatType: 'currency' },
  'costs_not_included'
]

export const lotteryInfoFields = [
  'lottery_date',
  'lottery_results_date',
  'lottery_venue',
  'lottery_status',
  'lottery_street_address',
  'lottery_summary',
  'lottery_city'
]

export const appInfoFields = [
  'application_phone',
  'office_hours',
  'application_organization',
  'application_street_address',
  'application_city',
  'application_state',
  'application_postal_code',
  { field: 'download_url', label: 'Download URL' }
]

export const agentDevInfoFields = [
  'leasing_agent_name',
  'leasing_agent_title',
  'leasing_agent_email',
  'leasing_agent_phone'
]

export const eligibilityRulesFields = [
  'building_selection_criteria',
  'eviction_history',
  'criminal_history',
  'credit_rating'
]

export const additionalInfoFields = [
  'required_documents',
  'smoking_policy',
  { field: 'legal_disclaimers', renderType: 'html' },
  'pet_policy'
]

export const openHousesFields = ['date', 'start_time', 'end_time']

export const infoSessionsFields = [
  'date',
  'start_time',
  'end_time',
  'venue',
  'street_address',
  'city'
]
