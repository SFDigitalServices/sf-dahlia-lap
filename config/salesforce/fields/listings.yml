listings:
  index_fields:
    Id:
    Name:
      label: 'Listing Name'
      minWidth: 225
    Application_Due_Date__c:
      type: 'date'
    Lottery_Date__c:
      type: 'date'
    Lottery_Results_Date__c:
      type: 'date'
    Lottery_Status__c:
    nFlagged_Applications__c:
      label: 'Flagged Applications'
    In_Lottery__c:
      label: 'Applications In Lottery'
  # detail_fields copied from:
  # https://github.com/Exygy/sf-dahlia-salesforce/blob/master/src/classes/HOP_ListingController.cls
  show_fields:
    Id:
    OwnerId:
    Owner.Name:
    Application_Due_Date__c:
    Name:
    Status__c:
    Building__c:
    Building__r.Name:
      label: 'Building'
    Min_BR__c:
    Lottery_Winners__c:
    Max_BR__c:
    Lottery_Results__c:
    Min_Income__c:
    Account__c:
    Account__r.Name:
    Max_Income__c:
    Min_Occupancy__c:
    Max_Occupancy__c:
    Building_Name__c:
    Neighborhood__c:
    Building_Street_Address__c:
    Developer__c:
    Building_City__c:
    Building_URL__c:
    Building_State__c:
    Year_Built__c:
    Building_Zip_Code__c:
    Description__c:
    Lottery_Preferences__c:
    Accessibility__c:
    Fee__c:
    Amenities__c:
    Deposit_Min__c:
    Costs_Not_Included__c:
    Deposit_Max__c:
    Lottery_Date__c:
    Lottery_Results_Date__c:
    Lottery_Venue__c:
    Lottery_Summary__c:
    Lottery_Street_Address__c:
    Lottery_City__c:
    Lottery_URL__c:
    Reserved_community_type__c:
    # ------------------
    Application_Phone__c:
    Application_Organization__c:
    Application_Street_Address__c:
    Application_City__c:
    Download_URL__c:
    Application_State__c:
    Application_Postal_Code__c:
    In_Lottery__c:
      label: 'Applications in Lottery'
    Leasing_Agent_Name__c:
    Leasing_Agent_Title__c:
    Leasing_Agent_Email__c:
    Leasing_Agent_Phone__c:
    Legal_Disclaimers__c:
    Building_Selection_Criteria__c:
    Pet_Policy__c:
    Report_id__c:
    Required_Documents__c:
    Smoking_Policy__c:
    Eviction_History__c:
    Criminal_History__c:
    Credit_Rating__c:
    Lottery_Status__c:
    Office_Hours__c:
    Tenure__c:
  lease_up_listing_fields:
    Id:
    Name:
    Lottery_Date__c:
    Lottery_Results_Date__c:
    In_Lottery__c:
    Units_Available__c:
    Lease_Signed_Application__c:
    LastModifiedDate:

