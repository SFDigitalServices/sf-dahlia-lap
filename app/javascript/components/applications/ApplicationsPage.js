import React from 'react'

import mapProps from 'utils/mapProps'

import TableLayout from '../layouts/TableLayout'
import { fetchApplications } from './actions'
import ApplicationsTableContainer from './ApplicationsTableContainer'

const ApplicationsPage = (props) => {
  const pageHeader = {
    title: 'Applications'
  }

  return (
    <TableLayout pageHeader={pageHeader}>
      <ApplicationsTableContainer {...props} />
    </TableLayout>
  )
}

const mapProperties = ({ listings }) => {
  return {
    onFetchData: fetchApplications,
    listings
  }
}

export default mapProps(mapProperties)(ApplicationsPage)
