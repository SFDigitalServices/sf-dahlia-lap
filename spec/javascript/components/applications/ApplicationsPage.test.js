import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import moment from 'moment'

import ApplicationsPage from 'components/applications/ApplicationsPage'
import ApplicationsTableContainer from 'components/applications/ApplicationsTableContainer'

import mockApplicationsPage from '../../fixtures/applications_page'
import listings from '../../fixtures/listings'

jest.mock('apiService', () => {
  return {
    fetchApplications: async () => {
      return { records: mockApplicationsPage, pages: 1 }
    }
  }
})

// FIXME: extract this out and make it generic
var wait = ms => new Promise((resolve, reject) => setTimeout(resolve, ms))

describe('ApplicationsPage', () => {
  test('should render successfully', (done) => {
    const wrapper = renderer.create(
      <ApplicationsPage />
    )

    expect(wrapper.toJSON()).toMatchSnapshot()
    wait(2000)
    done()
  })
  test('should display data correctly', async (done) => {
    const wrapper = mount(
      <ApplicationsPage listings={listings} />
    )
    await wait(2000)
    const table = wrapper.find(ApplicationsTableContainer)
    table.setState({ loading: false }, () => {
      const firstRowData = wrapper.find('div.rt-tr-group div.rt-tr').first()

      expect(firstRowData.text()).toContain(mockApplicationsPage[0].listing.name)
      expect(firstRowData.text()).toContain(moment(mockApplicationsPage[0].listing.lottery_date, '').format('MM/DD/YYYY'))
    })
    done()
  })
})
