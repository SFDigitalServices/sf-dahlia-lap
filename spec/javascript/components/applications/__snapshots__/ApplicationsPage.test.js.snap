// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationsPage should render succesfully 1`] = `
Array [
  <header
    className="lead-header short  bg-dust"
  >
    <div
      className="row full-width inner--3x"
    >
      <div
        className="large-12 columns"
      >
        <hgroup
          className="lead-header_group"
        >
          <h1
            className="lead-header_title small-serif c-oil"
          >
            Applications
          </h1>
        </hgroup>
      </div>
    </div>
  </header>,
  <section
    className="tabs-card-row row full-width inner--3x margin-bottom--2x padding-top--2x"
  >
    <div
      className="ReactTable"
      style={Object {}}
    >
      <div
        className="rt-table"
        style={undefined}
      >
        <div
          className="rt-thead -header"
          style={
            Object {
              "minWidth": "35px",
            }
          }
        >
          <div
            className="rt-tr"
            style={undefined}
          >
            <div
              className="rt-th"
              onClick={[Function]}
              role="heading"
              style={
                Object {
                  "flex": "35 0 auto",
                  "maxWidth": "35px",
                  "width": "35px",
                }
              }
            >
              <div
                className=""
              />
            </div>
          </div>
        </div>
        <div
          className="rt-tbody"
          style={
            Object {
              "minWidth": "35px",
            }
          }
        >
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -odd"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
          <div
            className="rt-tr-group"
          >
            <div
              className="rt-tr -padRow -even"
              style={Object {}}
            >
              <div
                className="rt-td"
                style={
                  Object {
                    "flex": "35 0 auto",
                    "maxWidth": "35px",
                    "width": "35px",
                  }
                }
              >
                <span>
                   
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="pagination-bottom"
      >
        <div
          className="-pagination"
          style={undefined}
        >
          <div
            className="-previous"
          >
            <button
              className="-btn"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              Previous
            </button>
          </div>
          <div
            className="-center"
          >
            <span
              className="-pageInfo"
            >
              Page
               
              <div
                className="-pageJump"
              >
                <input
                  onBlur={[Function]}
                  onChange={[Function]}
                  onKeyPress={[Function]}
                  type="number"
                  value={1}
                />
              </div>
               
              of
               
              <span
                className="-totalPages"
              >
                1
              </span>
            </span>
            <span
              className="select-wrap -pageSizeOptions"
            >
              <select
                onChange={[Function]}
                value={20}
              >
                <option
                  value={5}
                >
                  5
                   
                  rows
                </option>
                <option
                  value={10}
                >
                  10
                   
                  rows
                </option>
                <option
                  value={20}
                >
                  20
                   
                  rows
                </option>
                <option
                  value={25}
                >
                  25
                   
                  rows
                </option>
                <option
                  value={50}
                >
                  50
                   
                  rows
                </option>
                <option
                  value={100}
                >
                  100
                   
                  rows
                </option>
              </select>
            </span>
          </div>
          <div
            className="-next"
          >
            <button
              className="-btn"
              disabled={true}
              onClick={[Function]}
              type="button"
            >
              Next
            </button>
          </div>
        </div>
      </div>
      <div
        className="rt-noData"
      >
        No rows found
      </div>
      <div
        className="-loading"
      >
        <div
          className="-loading-inner"
        >
          Loading...
        </div>
      </div>
    </div>
  </section>,
]
`;
